apps/labs/posts/python-moa-introduction.md

```diff
@@ -0,0 +1,233 @@
+
+A = LazyArray(shape=('n', 'm'), name='A')
+B = LazyArray(shape=('k', 'l'), name='B')
+
+expression = (A + B).T.reduce('+')[0]
+expression
+```
+
+![Diagram of the python-moa compiler.](/posts/python-moa-introduction/frontend.png)
+
+## Shape Calculation
+
+The shape calculation is responsible for calculating the shape at every
+step of the computation. This means that operations have a shape. Note
+that the compiler handles symbolic shapes thus the exact shape does not
+need to be known, only the dimension. After the shape calculation step
+we can guarantee that an algorithm is a valid program and there will be
+no out of bound memory accesses. Making MOA extremely compelling for
+[FPGAs](https://en.wikipedia.org/wiki/Field-programmable_gate_array) and
+compute units with a minimal OS. If an algorithm makes it past this
+stage and fails then it is an issue with the compiler not the algorithm.
+
+```python
+expression.visualize(stage='shape')
+```
+
+![Diagram of shape calculation.](/posts/python-moa-introduction/shape_calculation.png)
+
+## Denotational Normal Form (DNF)
+
+The DNF\'s responsibility is to reduce the high level MOA expression to
+the minimal and optimal machine independent computation. This graph has
+all of the indexing patterns of the computation and resulting shapes.
+Notice that several operations disappear in this stage such a transpose
+because transpose is simply index manipulation.
+
+```python
+expression.visualize(stage='dnf')
+```
+
+![Diagram of Denotational Normal Form.](/posts/python-moa-introduction/denotational_normal_form.png)
+
+## Operational Normal Form (ONF)
+
+The ONF is the stage of the compiler that will have to be the most
+flexible. At its current stage the ONF is a naive compiler that does not
+perform many important optimizations such as [PSI
+reduction](https://www.researchgate.net/publication/264758384_Effective_data_parallel_computation_using_the_Psi_calculus)
+which reduces the number of loops in the calculation, loop ordering, and
+minimize the number of accumulators. MOA has ideas of dimension lifting
+which make parallization and optimizing for cache sizes much easier.
+Additionally algorithms must be implemented differently for sparse,
+column major, row major. The ONF stage is responsible for any
+\"optimal\" machine dependent implementation. \"optimal\" will vary from
+user to user and thus will have to allow for multiple programs: optimal
+single core, optimal parallel, optimal gpu, optimal low memory, etc.
+
+```python
+print(expression.compile(use_numba=True, include_conditions=False))
+```
+
+    @numba.jit
+    def f(A, B):
+
+
+        n = A.shape[0]
+
+        m = A.shape[1]
+
+        k = B.shape[0]
+
+        l = B.shape[1]
+
+        _a21 = numpy.zeros(())
+
+        _a19 = numpy.zeros(())
+
+        _a21 = 0
+
+        for _i10 in range(0, m, 1):
+
+            _a21 = (_a21 + (A[(0, _i10)] + B[(0, _i10)]))
+
+        _a19[()] = _a21
+        return _a19
```

Comment:
[Indented code does not work in MDX](https://mdxjs.com/docs/what-is-mdx/#markdown).

Also, this looks like output to me, so we have the same issue here that we have described before:

- https://github.com/Quansight/Quansight-website/pull/647#pullrequestreview-1279809159